rasa/core/story_conflict.py

```diff
@@ -0,0 +1,252 @@
+
+from rasa.core.actions.action import ACTION_LISTEN_NAME
+from rasa.core.domain import PREV_PREFIX, Domain
+from rasa.core.events import ActionExecuted, Event
+from rasa.core.featurizers import MaxHistoryTrackerFeaturizer
+from rasa.nlu.constants import MESSAGE_INTENT_ATTRIBUTE
+from rasa.core.training.generator import TrackerWithCachedStates
+
+
+class StoryConflict:
+    def __init__(
+        self, sliced_states: List[Optional[Dict[Text, float]]],
+    ):
+        self.sliced_states = sliced_states
+        self.hash = hash(str(list(sliced_states)))
+        self._conflicting_actions = {}  # {"action": ["story_1", ...], ...}
+        self.correct_response = None
+
+    @staticmethod
+    def find_conflicts(
+        trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+    ) -> List:
+        """
+        Generate a list of StoryConflict objects, describing
+        conflicts in the given trackers.
+        :param trackers: Trackers in which to search for conflicts
+        :param domain: The domain
+        :param max_history: The maximum history length to be
+        taken into account
+        :return: List of conflicts
+        """
+
+        # We do this in two steps, to reduce memory consumption:
+
+        # Create a 'state -> list of actions' dict, where the state is
+        # represented by its hash
+        rules = StoryConflict._find_conflicting_states(trackers, domain, max_history)
+
+        # Iterate once more over all states and note the (unhashed) state,
+        # for which a conflict occurs
+        conflicts = StoryConflict._build_conflicts_from_states(
+            trackers, domain, max_history, rules
+        )
+
+        return conflicts
+
+    @staticmethod
+    def _find_conflicting_states(
+        trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+    ) -> Dict[int, Optional[List[Text]]]:
+        # Create a 'state -> list of actions' dict, where the state is
+        # represented by its hash
+        rules = {}
+        for tracker, event, sliced_states in StoryConflict._sliced_states_iterator(
+            trackers, domain, max_history
+        ):
+            h = hash(str(list(sliced_states)))
+            if h not in rules:
+                rules[h] = [event.as_story_string()]
+            elif h in rules and event.as_story_string() not in rules[h]:
+                rules[h] += [event.as_story_string()]
+
+        # Keep only conflicting rules
+        return {
+            state_hash: actions
+            for (state_hash, actions) in rules.items()
+            if len(actions) > 1
+        }
+
+    @staticmethod
+    def _build_conflicts_from_states(
+        trackers: List["TrackerWithCachedStates"],
+        domain: Domain,
+        max_history: int,
+        rules: Dict[int, Optional[List[Text]]],
+    ):
+        # Iterate once more over all states and note the (unhashed) state,
+        # for which a conflict occurs
+        conflicts = {}
```

Comment:
same as above. You could use a `defaultdict`.

Reply:
How? The conflict values are classes (`StoryConflict`s) that need to be created with an argument `sliced_states`, which is not available at this point.